### ISSUE TITLE
`Invalid argument - clock_gettime (Errno::EINVAL)` error thrown by profiler (now and then)

### ISSUE THREAD

**Issue by matstc** [DataDog/dd-trace-rb#1876]
We have cron jobs set up that are using the DD profiler. Sometimes a job will end abruptly and the stack trace below is printed.

Could anyone shed some light onto what this code is trying to do? The error EINVAL seems to indicate an error calling a POSIX function (?)

Does this indicate a bug in the profiler, or in the underlying native code?

Any help would be appreciated.

```
#<Thread:0x000055e4555af110@Datadog::Profiling::Collectors::Stack@/app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/ext/cthread.rb:83 run> terminated with exception (report_on_exception is true):
/app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/ext/cthread.rb:95:in `clock_gettime': Invalid argument - clock_gettime (Errno::EINVAL)
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/ext/cthread.rb:95:in `cpu_time'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:155:in `get_cpu_time_interval!'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:131:in `collect_thread_event'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:108:in `block in collect_events'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:104:in `each'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:104:in `collect_events'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:92:in `block in collect_and_wait'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/utils/time.rb:39:in `measure'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:91:in `collect_and_wait'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:83:in `perform'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/loop.rb:18:in `block in perform'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/loop.rb:108:in `block in perform_loop'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/loop.rb:104:in `loop'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/loop.rb:104:in `perform_loop'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/loop.rb:18:in `perform'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/async.rb:22:in `block in perform'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/async.rb:135:in `block in start_worker'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/ext/cthread.rb:86:in `block in initialize'
/app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/ext/cthread.rb:95:in `clock_gettime': Invalid argument - clock_gettime (Errno::EINVAL)
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/ext/cthread.rb:95:in `cpu_time'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:155:in `get_cpu_time_interval!'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:131:in `collect_thread_event'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:108:in `block in collect_events'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:104:in `each'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:104:in `collect_events'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:92:in `block in collect_and_wait'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/utils/time.rb:39:in `measure'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:91:in `collect_and_wait'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/collectors/stack.rb:83:in `perform'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/loop.rb:18:in `block in perform'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/loop.rb:108:in `block in perform_loop'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/loop.rb:104:in `loop'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/loop.rb:104:in `perform_loop'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/loop.rb:18:in `perform'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/async.rb:22:in `block in perform'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/workers/async.rb:135:in `block in start_worker'
        from /app/out/ruby/2.5.0/gems/ddtrace-0.53.0/lib/ddtrace/profiling/ext/cthread.rb:86:in `block in initialize'
```

**Comment by ivoanjo** [comment]
Hey @matstc, thanks for reporting this, and sorry that you ran into the issue!

The TL;DR version is: this should be fixed in the latest ddtrace release (v0.54.2 as of this writing) as we're explicitly handling that exception now:

https://github.com/DataDog/dd-trace-rb/blob/6b0b1d51a4a845339a6b986841f66077034c3e6b/lib/ddtrace/profiling/native_extension.rb#L32-L36

The non-TL;DR version is:

1. The [Linux manual pages](https://linux.die.net/man/3/clock_gettime) are somewhat vague about when `EINVAL` is returned. We've found that one case this happens is when the thread being profiled dies while the profiler is was looking at it. 

2. The issue you mention should affect only the profiler, and stop it. It should not otherwise impact your application, since that's just a background component (the profiler thread) that dies.

    Your message hinted at your application being affected -- I'd be curious to know more. I wonder if some other thread in your workload is dieing abruptly, and does so while the profiler is looking at it, thus also taking the profiler with it. We don't expect the reverse to happen, but again, if this is what you're seeing, we'll definitely want to know and look into it! :)

**Comment by matstc** [comment]
Thank you for the detailed answer @ivoanjo!

My own bit of research also didn't enlighten me as to what `EINVAL` really means. I'm afraid I can't recommend much more than what you are already doing (i.e. `rescue nil`).

You are right, our background job is likely not affected by this. It likely just ends around the same time as that erroring thread because it's short enough.

I will keep the gem up to date and come back here if I see any more issues related to this.

**Comment by ivoanjo** [comment]
It's been a few days, and I'm hoping no news is good news!

I'm going to go ahead and close this issue, but do feel free to reopen and report any issues or wonky things we can make better :)

**Comment by ariccio** [comment]
This is very bizarre. I'm seeing this for the first time ever just now on Windows, [here](https://github.com/heroku/barnes/blob/main/lib/barnes/instruments/stopwatch.rb#L77) in Barnes. I don't even see a codepath that's changed recently that could throw EINVAL.

**Comment by ivoanjo** [comment]
Hey @ariccio, not sure I can help you there, since barnes is not related to dd-trace-rb 😅

I suspect a similar solution to what was done in dd-trace-rb would work there: correctly handle the exception and either retry or accept that sometimes there's no value to be reported 😃

**Comment by ariccio** [comment]
I don't really expect much! I just find that by adding info, and also often from linking related issues together on GitHub, it's helpful for whoever _may_ figure it out down the line. A bit of detective work that can be shared, y'know?

**Comment by ariccio** [comment]
Strangely, for example, it seemingly went away for me? I don't remember what I did, if anything, but it didn't involve patching anything.... Which sounds to me like something isn't actually working the way we expect!
